i18n first commit

k-nakamura1984 opened the pull request.

k-nakamura1984 committed:
i18n first commit

k-nakamura1984 commented on app/models/book.rb:
```diff
@@ -1,5 +1,7 @@
 # frozen_string_literal: true
 
 class Book < ApplicationRecord
+  validates :title, presence: true
+  validates :memo, presence: true
```
エラーメッセージが翻訳されるかどうか確認のために記述しました。

k-nakamura1984 commented on app/views/books/edit.html.erb:
```diff
@@ -1,6 +1,6 @@
 
-<%= link_to 'Show', @book %> |
-<%= link_to 'Back', books_path %>
+<%= link_to t('button.show'), @book %> |
+<%= link_to t('button.back'), books_path %>
```
buttonがどこを参照しているのか、よくわからなかったです…。

yoshitsugu replied:
基本的にyamlの階層構造をそのまま適用されています
https://railsguides.jp/i18n.html#%E7%8B%AC%E8%87%AA%E3%81%AE%E8%A8%B3%E6%96%87%E3%82%92%E4%BF%9D%E5%AD%98%E3%81%99%E3%82%8B%E6%96%B9%E6%B3%95

k-nakamura1984 replied:
自分でcommon.ja.ymlファイルに記述しておきながら、ymlファイルのキーが他の箇所を参照していると錯誤しておりました…。

tがi18nのメソッドであり、引数はymlに設定したキーであることに注意したいと思います。

k-nakamura1984 commented on app/views/books/index.html.erb:
```diff
@@ -1,14 +1,14 @@
-      <th>Memo</th>
-      <th>Author</th>
-      <th>Picture</th>
+      <th><%= t :title%></th>
+      <th><%= t :memo%></th>
+      <th><%= t :author%></th>
+      <th><%= t :picture%></th>
```
en.ymlに記述しなくても、locale=enを指定した際に、Title, Memoといったように翻訳される挙動がよくわからなかったです。

yoshitsugu replied:
僕も詳しくは追ってないですが、単語だけだったらそのままシンボルの値をcapitalizeしたものが使われるみたいですね。

yoshitsugu replied:
ここは、モデルの属性名なので、 `human_attribute_name` を使うと便利です
https://railsguides.jp/i18n.html#active-record%E3%83%A2%E3%83%87%E3%83%AB%E3%81%A7%E7%BF%BB%E8%A8%B3%E3%82%92%E8%A1%8C%E3%81%AA%E3%81%86

k-nakamura1984 replied:
> 単語だけだったらそのままシンボルの値をcapitalizeしたものが使われるみたいですね。

仕様ということで理解しました！

>モデルの属性名なので、 human_attribute_name を使うと便利です

どのように便利なのか、いまいち理解できませんでした。
参考情報を再読したのですが、本件は下記の認識でよろしいでしょうか？
- `t :title` => `Book.human_attribute_name("title")`
-  ymlファイルをmodel用に分けなくて良くなる。ただしja.ymlファイルに定義する必要はある。

yoshitsugu replied:
@k-nakamura1984 
> ymlファイルをmodel用に分けなくて良くなる。ただしja.ymlファイルに定義する必要はある。

いえ、どちらの場合でもyamlファイルの設計は柔軟にできます。
訳を書く際のyamlの構造が違うって感じですね。`human_attribute_name` が参照するのは、 `(言語).activerecord.attributes.(モデル名).(属性名)` のような構造の訳になります。 
この場合のメリットとしては、例えば、

- Bookというモデルがあって、 `title` カラムを持つとき `題名` と訳したいとします。
- 他にMusicというモデルがあって `title` カラムを持ち、こちらは `曲名` と訳したいとします。

このような場合、 `t('title')` で `title` と訳語が1対1対応している場合だと、コンテキストによる訳の変更に対応できません。
`human_attribute_name` をつかうと、上記の通りモデルごとに訳をわけられるので、このような場合も対応可能です。

k-nakamura1984 replied:
確認ありがとうございます。
ymlファイルの使い分けや、human_attribute_nameの使い方や便利な点について、理解できた気がします。

> t('title') で title と訳語が1対1対応している場合だと、コンテキストによる訳の変更に対応できません。

例えば、上記の場合`activerecord.attribute.Book.title``activerecord.attribute.Musit.title`とそれぞれ書くことで訳し分けるはできる。しかしhuman_attribute_nameを用いることで訳の回想を確認せずとも、モデルごとにキーを訳し分けることができる、という理解はあっておりますでしょうか？

yoshitsugu replied:
@k-nakamura1984 そうですね。そんな感じです。

yoshitsugu replied:
@k-nakamura1984 ですので、ここも `human_attribute_name` 使ってみてください〜

k-nakamura1984 commented on config/locales/common.ja.yml:
```diff
@@ -0,0 +1,8 @@
+ja:
+  button:
+    edit: 編集
+    back: 戻る
+    show: 詳細
+    index: 一覧
+    destroy: 削除
+    new book: 新規作成
```
「独習Ruby on Rails」を参照して、ファイルを分けました。
同様の内容をja.ymlに記載して、本ファイルをブランクにした際にエラーが出る理由がよくわからなかったです…。
別途作成したmodel.ja.ymlなどのように特定の名前でymlファイルを作成すると、そちらが優先されるような仕組みなのでしょうか。

k-nakamura1984 replied:
別途コメントいただいた`human_attribute_name`の利用により解消されました！

k-nakamura1984 commented on config/locales/en.yml:
```diff
@@ -30,4 +30,18 @@
-  hello: "Hello world"
+  'Are you sure?': "Are you sure?"
+
+  books:
+    create:
+      "Book was successfully created": "Book was successfully created."
```
左側（キー？）の末尾を`.`にすると、シンタックスエラーが表示される理由を見つけることができませんでした…。

yoshitsugu replied:
シンタックスエラーになるんですね？ :thinking: 
なんでだろう。
そもそもあまりこういうキーが文章になるような形は見かけないんですが、
https://railsguides.jp/i18n.html#%E8%A8%B3%E6%96%87%E3%82%92%E4%B8%8E%E3%81%88%E3%81%A6%E6%96%87%E5%AD%97%E5%88%97%E3%82%92%E5%9B%BD%E9%9A%9B%E5%8C%96%E3%81%99%E3%82%8B
ここの例のようにスネークケースにすることはできなさそうでしょうか？

yoshitsugu replied:
また、対訳をもっと一般化するために、変数を使う方法などもあるので、参考にしてみてください〜
https://railsguides.jp/i18n.html#%E8%A8%B3%E6%96%87%E3%81%AB%E5%A4%89%E6%95%B0%E3%82%92%E6%B8%A1%E3%81%99

k-nakamura1984 commented on config/locales/en.yml:
```diff
@@ -30,4 +30,18 @@
 # available at https://guides.rubyonrails.org/i18n.html.
 
 en:
```
特に言語を指定しなくても英語に翻訳される箇所と、en.ymlを指定しないと翻訳されない（missing translationになる）の違いがよくわからなかったです。

yoshitsugu replied:
ちょっと調べてみたんですが、viewから呼ぶとHelperが使われるので、
https://api.rubyonrails.org/v6.1.3.2/classes/ActionView/Helpers/TranslationHelper.html#method-i-translate
こちらがよばれて、デフォルトで翻訳のキーの最後のセグメントが表示されるみたいです

k-nakamura1984 replied:
すみません…。いただいたコメントの理解が難しかったです。
>こちらがよばれて、

「こちら」が具体的にどの箇所を指すのか特定できませんでした。

>デフォルトで翻訳のキーの最後のセグメントが表示されるみたいです

「最後のセグメント」が何を指すのか特定できませんでした。

お手数おかけし申し訳ないのですが、もう少し詳しく教えていただけますでしょうか。

yoshitsugu replied:
@k-nakamura1984 
すいません、紛らわしかったですね！
こちら、で指していたのはリンク先のheperのメソッドでした。
最後のセグメントとは `xxx.yyy.zzz` で参照する場合 `zzz` の部分、ということです。

実際、意図してここを使いわけることは少ないので、普通は全部訳のファイルを用意することが多いです。
なので、あまり意識しなくてもよいかなーと思います :ok_man:

k-nakamura1984 replied:
再度教えていただき、ありがとうございます。
申し訳ないのですが、いただいたリンク先の内容を要約できる程度に理解することできませんでした。リンク先の大まかな内容はtranslation missingが起きる時の条件と、起きた場合にどの値を返すか、ということが記載されているという理解であっておりますでしょうか？

>デフォルトで翻訳のキーの最後のセグメントが表示されるみたいです

上記が、いただいたリンク先の情報のどの箇所から導かれれるのか見当をつけることができませんでした。読解力が不足しておりすみません…。

>en.ymlを指定しないと翻訳されない（missing translationになる）の違いがよくわからなかった

という私の不明点については、en.ymlに記載しなくてもヘルパーの処理によって、最後のセグメントが表示されて、特に翻訳を定義していなくても翻訳されたように見える場合がある、と理解しました。

yoshitsugu replied:
@k-nakamura1984 あ、ここではコードを理解する必要はないですよ〜。

> en.ymlに記載しなくてもヘルパーの処理によって、最後のセグメントが表示されて、特に翻訳を定義していなくても翻訳されたように見える場合がある、と理解しました。

これが理解できていれば十分です！

k-nakamura1984 commented on config/locales/ja.yml:
```diff
@@ -0,0 +1,227 @@
+ja:
```
22line以降は、「独習Ruby on Rails」を参照して、Githubの標準の辞書を反映しました。
[rails\-i18n/ja\.yml at master · svenfuchs/rails\-i18n](https://github.com/svenfuchs/rails-i18n/blob/master/rails/locale/ja.yml)

yoshitsugu reviewed (commented):
コメントしました。
Railsガイドのi18nのページ読むと理解が深まるところもあるかもしれません！
https://railsguides.jp/i18n.html

yoshitsugu commented on config/locales/en.yml:
```diff
@@ -30,4 +30,18 @@
+    create:
+      "Book was successfully created": "Book was successfully created."
+    update:
+      "Book was successfully updated": "Book was successfully updated."
+    destroy:
+      "Book was successfully destroyed": "Book was successfully destroyed."
```
ここも文章ではなくて変数などを使ってできるだけ汎用的に書いておきたいですね。
`YAGNI(You ain't gonna need it)` という言葉もあり、汎用的なのを求めすぎるのも考えものですが、ここは今後も `XXX was successfully updated` みたいなのがたくさん必要になることは容易に想像できるので、汎用化をやってみてください〜。

yoshitsugu reviewed (commented):
コメントしました！

k-nakamura1984 committed:
i18nのプラクティスの課題提出のため、下記を修正

model_name.human、human_attribute_nameに置き換え。
model_ja.yml, common_ja.ymlを削除して、ja.ymlに統合。
長文の翻訳キーを変数に修正。

k-nakamura1984 commented on app/controllers/books_controller.rb:
```diff
@@ -28,7 +28,7 @@ def create
     respond_to do |format|
       if @book.save
-        format.html { redirect_to @book, notice: t('.Book was successfully created.') }
+        format.html { redirect_to @book, notice: t('notice.create', title: @book.title) }
```
変数を使って汎用化を試みました。
Bookモデル以外に利用できないため、汎用化しきれているか自信がありません。
いかがでしょうか？

k-nakamura1984 commented on app/views/books/show.html.erb:
```diff
@@ -1,10 +1,5 @@
-  <strong>Title:</strong>
-  <%= @book.title %>
-</p>
-
```
重複箇所を削除しました。

yoshitsugu reviewed (approved):
かなりスッキリしていい感じだと思います！LGTMです！
